python/pywatchman/__init__.py

```diff
@@ -114,16 +126,128 @@ def __init__(self):
 
     LocalFree = ctypes.windll.kernel32.LocalFree
 
-    GetOverlappedResultEx = ctypes.windll.kernel32.GetOverlappedResultEx
-    GetOverlappedResultEx.argtypes = [wintypes.HANDLE,
-                                      ctypes.POINTER(OVERLAPPED), LPDWORD,
-                                      wintypes.DWORD, wintypes.BOOL]
-    GetOverlappedResultEx.restype = wintypes.BOOL
+    # Windows 7 and earlier does not support GetOverlappedResultEx
+    # The alternative is to use GetOverlappedResult and wait
+    # for read or write operation to complete. This is
+    # done be using CreateEvent and WaitForSingleObject.
+    # Windows 8 is version 6.2 - see the Operating System Version
+    # page on MSDN
+    WINDOWS_8_AND_LATER = sys.getwindowsversion() >= (6, 2)
+    if not WINDOWS_8_AND_LATER:
+        CreateEvent = ctypes.windll.kernel32.CreateEventA
+        CreateEvent.argtypes = [LPDWORD, wintypes.BOOL, wintypes.BOOL,
+                                wintypes.LPSTR]
+        CreateEvent.restype = wintypes.HANDLE
+
+        GetOverlappedResult = ctypes.windll.kernel32.GetOverlappedResult
+        GetOverlappedResult.argtypes = [wintypes.HANDLE,
+                                        ctypes.POINTER(OVERLAPPED), LPDWORD,
+                                        wintypes.BOOL]
+        GetOverlappedResult.restype = wintypes.BOOL
+
+        WaitForSingleObject = ctypes.windll.kernel32.WaitForSingleObject
+        WaitForSingleObject.argtypes = [wintypes.HANDLE, wintypes.DWORD]
+        WaitForSingleObject.restype = wintypes.DWORD
+    else:
+        GetOverlappedResultEx = ctypes.windll.kernel32.GetOverlappedResultEx
+        GetOverlappedResultEx.argtypes = [wintypes.HANDLE,
+                                          ctypes.POINTER(OVERLAPPED), LPDWORD,
+                                          wintypes.DWORD, wintypes.BOOL]
+        GetOverlappedResultEx.restype = wintypes.BOOL
```

Comment:
Thanks for being so thorough!  The implementation in the service was a bit speculative, since I didn't have access to a system running Windows 7, so it is quite possible that it is faulty.

First: good catch.  We should fix this in the C code too.

Second: hmm, what if we changed the `if durationmillis <= 0:` condition to be just `if durationmillis < 0:`?  That way we'd do another iteration at the end.

Third: Interesting.  Does it have to be a 10ms sleep?  Does a 0ms sleep have the same positive effect?   This probably also affects the C code and I'd love to have a solution in both areas.

We're not really geared up for debugging Windows code here and I don't have a lot of bandwidth to get hands on... are you able to build and run watchman's tests for yourself?  It may just be as simple as running `make` and then running `make integration` (there's a `make.bat` script in the top level that drives this).  It should be pretty self-contained; I think the only mandatory dependency is PHP for some of the legacy tests.

If you think it is easier to manage this using an Event, and can find a way to get that working without adding a lot of conditional code, I'd be happy to see that land!